mdbenchmark/plot.py

```diff
@@ -0,0 +1,185 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# MDBenchmark is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with MDBenchmark.  If not, see <http://www.gnu.org/licenses/>.
+import pandas as pd
+import matplotlib.pyplot as plt
+from matplotlib.backends.backend_agg import FigureCanvasAgg as FigureCanvas
+from matplotlib.figure import Figure
+import click
+
+from .cli import cli
+from .utils import calc_slope_intercept, lin_func, generate_output_name
+from . import console
+
+
+def plot_line(df, df_sel, label, ax=None):
+    if ax is None:
+        ax = plt.gca()
+
+    p = ax.plot(df_sel, 'ns/day', '.-', data=df, ms='10', label=label)
+    color = p[0].get_color()
+    slope, intercept = calc_slope_intercept(
+        (df[df_sel].iloc[0], df['ns/day'].iloc[0]), (df[df_sel].iloc[1],
+                                                     df['ns/day'].iloc[1]))
+    # avoid a label and use values instead of pd.Series
+    ax.plot(
+        df[df_sel],
+        lin_func(df[df_sel].values, slope, intercept),
+        ls='--',
+        color=color,
+        alpha=.5)
+    return ax
+
+
+def plot_over_group(df, plot_cores, ax=None):
+    # plot all lines
+    df_sel = 'ncores' if plot_cores else 'nodes'
+
+    gb = df.groupby(['gpu', 'module', 'host'])
+    groupby = ['gpu', 'module', 'host']
+    for key, df in gb:
+        label = ' '.join(['{}={}'.format(n, v) for n, v in zip(groupby, key)])
+        plot_line(df=df, df_sel=df_sel, ax=ax, label=label)
+
+    # style axes
+    xlabel = 'cores' if plot_cores else 'nodes'
+    ax.set_xlabel('Number of {}'.format(xlabel))
+    ax.set_ylabel('Performance [ns/day]')
+
+    lgd = ax.legend(loc='upper center', bbox_to_anchor=(0.5, -0.175))
+    plt.tight_layout()
+    # here I return the figure as well as the legend
+    return ax, lgd
+
+
+def filter_dataframe_for_plotting(df, host_name, module_name, gpu, cpu):
+    # gpu/cpu can be plotted together or separately
+    if gpu and cpu:
+        # if no flags are given by the user or both are set everything is plotted
+        console.info("Plotting GPU and CPU data.")
+    elif gpu and not cpu:
+        df = df[df.gpu]
+        console.info("Plotting GPU data only.")
+    elif cpu and not gpu:
+        df = df[~df.gpu]
+        console.info("Plotting CPU data only.")
+    elif not cpu and not gpu:
+        console.error("CPU and GPU not set. Nothing to plot. Exiting.")
+
+    for host in host_name:
+        if host in df['host'].tolist():
+            df = df[df['host'].isin(host_name)]
+            console.info('Data for the following hosts will be plotted: {}.', set(host_name))
```

Comment:
We should change the whole logic here. I'd propose the following sketch:

1) If no `hostname` is given, we plot all available hosts.
2) If `hostname` is given, we check the names for availability. This results in two cases:
a) One or more provided hostnames can not be found: we use `console.error` to notify the user and exit here.
b) All provided hostnames are available in the CSV/log files: we plot all given hosts.


This would change the code to this:

```python
# Case 1
if not host_name:
    console.info("Plotting all hosts in csv.")

df_filtered_hosts = df[df['hostname'].isin(host_name)]
df_unique_hosts = np.unique(df_filtered_hosts)

# Case 2a
if df_unique_hosts.size != len(host_name):
    console.error("Could not find all provided hosts. Available hosts are: XYZ")

# Case 2b
console.info("Data for the following hosts will be plotted: {}".format(df_unique_hosts))
```